src/passes/functionPruner/functionRemover.ts

```diff
@@ -15,24 +16,31 @@ export class FunctionRemover extends ASTMapper {
     this.reachableFunctions = new Set();
   }
 
+  importFuncFilter = (funcs: readonly FunctionDefinition[]) =>
+    funcs
+      .filter((func) => !this.reachableFunctions.has(func.id))
+      .filter(
+        (func) =>
+          !(
+            func instanceof CairoImportFunctionDefinition &&
+            func.functionStubKind === FunctionStubKind.TraitStructDefStub
+          ),
+      );
+
   visitSourceUnit(node: SourceUnit, ast: AST): void {
     node.vFunctions.filter((func) => isExternallyVisible(func)).forEach((func) => this.dfs(func));
 
     node.vContracts.forEach((c) => this.visitContractDefinition(c, ast));
 
-    node.vFunctions
-      .filter((func) => !this.reachableFunctions.has(func.id))
-      .forEach((func) => node.removeChild(func));
+    this.importFuncFilter(node.vFunctions).forEach((func) => node.removeChild(func));
```

Comment:
I guess, it'd make the difference in the dfs graph visitor, what if a function inside anyone of the contracts makes a call to the source unit level function.

Reply:
Source Unit functions are supposed to be file-level function definitions (free functions). I believe, currently those function get inline in the contracts, but I'm not sure if they are removed then. Anyway, there is still the question whether to remove them or not given that you can use them in Cairo 1, so I wouldn't remove this code yet.
I agree with @rjnrohit that we can not separate the order of operations that way: the dfs has to start with the free functions, but this will only visit the called functions inside their bodies, so it is still necessary to check if there is any function inside the contracts that is using those free functions, hence, the contracts in the source unit are visited. After that, is when the functions can be removed.